### ISSUE TITLE
Add a `as_constant` helper for forcing computations to compile-time

### ISSUE THREAD

**Issue by eric-wieser** [numba/numba#5339]
## Feature request

I'd like to be able to write something like the following:
```python
from types import SimpleNamespace
a = SimpleNamespace(b=2)

@numba.njit
def foo():
    # return a.b   # fails, because numba tries to evaluate at runtime
    return numba.as_constant(lambda: a.b)
```
My understanding is that `as_constant(a.b)` would be tricky to extract in numba, but I think `as_constant(lambda: a.b)` would be more easy to pull out the of the python bytecode.

A more complex situation would be one where I want the constant to depend on the argument types:
```python
def make_multiply_by(a):
    @numba.njit
    def inner(b):
        # trickier - want to expose the type information of `b` here, without actually using the value
        # information.
        a_copy = numba.as_constant(lambda: a.astype(b.dtype))
    return inner
```

**Comment by stuartarchibald** [comment]
> ## Feature request
> 
> I'd like to be able to write something like the following:
> 
> ```python
> from types import SimpleNamespace
> a = SimpleNamespace(b=2)
> 
> @numba.njit
> def foo():
>     # return a.b   # fails, because numba tries to evaluate at runtime
>     return numba.as_constant(lambda: a.b)
> ```

This is possible with restrictions. The `lambda` requirement does make it a lot easier to track for this case.

> My understanding is that `as_constant(a.b)` would be tricky to extract in numba, but I think `as_constant(lambda: a.b)` would be more easy to pull out the of the python bytecode.
> 
> A more complex situation would be one where I want the constant to depend on the argument types:
> 
> ```python
> def make_multiply_by(a):
>     @numba.njit
>     def inner(b):
>         # trickier - want to expose the type information of `b` here, without actually using the value
>         # information.
>         a_copy = numba.as_constant(lambda: a.astype(b.dtype))
>     return inner
> ```
What would `a` be here? A NumPy array that needs to be a constant but of a type determined by `b`'s dtype at compile time?

**Comment by eric-wieser** [comment]
> What would a be here? A NumPy array that needs to be a constant but of a type determined by b's dtype at compile time?

Yes, exactly. I'd be happy if in that lambda, `b` is bound to the numba type, and I have to do the dance to convert between numba and numpy types

**Comment by stuartarchibald** [comment]
> > What would a be here? A NumPy array that needs to be a constant but of a type determined by b's dtype at compile time?
> 
> Yes, exactly. I'd be happy if in that lambda, `b` is bound to the numba type, and I have to do the dance to convert between numba and numpy types

So something like:
```python
from numba.np.numpy_support import as_dtype

def make_multiply_by(a):
    @njit(pipeline_class=MyCompiler)
    def inner(g):
        def foo():
            # Know Numba type information for 'g'
            # Know actual object instance for 'a'
            return a.astype(as_dtype(g.dtype))
        a_copy = constexpr(foo)
        return a_copy
    return inner

z = make_multiply_by(np.array([3-2j, 4+7j]))
print(z(np.ones(3,)))
```
?

**Comment by eric-wieser** [comment]
Exactly

**Comment by eric-wieser** [comment]
Not sure of the relevance of `pipeline_class=MyCompiler` - I'd like this to work by default.

**Comment by stuartarchibald** [comment]
> Not sure of the relevance of `pipeline_class=MyCompiler` - I'd like this to work by default.

When developing compiler passes its often convenient to create a [new basic compiler and add the new pass to it](https://numba.readthedocs.io/en/stable/developer/custom_pipeline.html), means there's far less to debug, I left that in when copy pasting!
